src/runtime/hexagon/hexagon_device_api.cc

```diff
@@ -115,13 +120,34 @@ void* HexagonDeviceAPI::AllocDataSpace(Device dev, size_t nbytes, size_t alignme
   if (alignment < kHexagonAllocAlignment) {
     alignment = kHexagonAllocAlignment;
   }
-  return mgr->AllocateHexagonBuffer(nbytes, alignment, String("global"));
+  CHECK(runtime_hexbuffs) << "Attempted to allocate Hexagon data with "
+                          << "HexagonDeviceAPI::AllocDataSpace before initializing resources.  "
+                          << "Please call HexagonDeviceAPI::AcquireResources";
+  return runtime_hexbuffs->AllocateHexagonBuffer(nbytes, alignment, String("global"));
 }
 
 void HexagonDeviceAPI::FreeDataSpace(Device dev, void* ptr) {
   CHECK(ptr) << "buffer pointer is null";
   CHECK(IsValidDevice(dev)) << "dev.device_type: " << dev.device_type;
-  mgr->FreeHexagonBuffer(ptr);
+  if (runtime_hexbuffs) {
+    runtime_hexbuffs->FreeHexagonBuffer(ptr);
+  } else {
```

Comment:
Not sure on the value-add for this `else` case.  The only thing that happens is that we might fail in the `CHECK` if we free a buffer after we release resources that was not in the `released_runtime_buffers`.  Do we want this behavior?  What happens on-system if we just get rid of this `else` case?  We can then get rid of `released_runtime_buffers` and the need for `HexagonBufferManager::current_allocations`.  We can simply ...

```
  if (runtime_hexbuffs) {
    runtime_hexbuffs->FreeHexagonBuffer(ptr);
  }
```

With some comments to explain why the "free" may happen after the "release".

Reply:
My previous reply is related.

If there's a manager, I can just use it to call Free.  If there isn't (it has been released, or maybe acquire was never called), I can check what I had allocated before Release.  If that buffer is in there, I will not throw.

Reply:
Making sure all "temporal" frees occur before Release is a good idea and worthwhile endeavor. ✅ 

I am not convinced that two `HexagonBufferManager` objects are required. Instead, you could have separate interfaces in a single `HexagonBufferManager` class to track "permanent" (for RPC e.g.) versus "temporal" allocations.  And, in this way, alleviate the need for the `current_allocations` method which exposes the internals of the `HexagonBufferManager` class to its users.  Something to consider for a future cleanup, perhaps.